pkg/logger/progress_logger.go

```diff
@@ -0,0 +1,130 @@
+/*
+Copyright 2022 VMware, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package logger
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	pb "github.com/cheggaaa/pb/v3"
+	"github.com/go-logr/logr"
+	regv1 "github.com/google/go-containerregistry/pkg/v1"
+)
+
+// ProgressLogger Progress bar
+type ProgressLogger interface {
+	Start(ctx context.Context, progress <-chan regv1.Update)
+	End()
+}
+
+// NewProgressBar constructor to build a ProgressLogger responsible for printing out a progress bar using updates when
+// writing to a registry via ggcr
+func NewProgressBar(logger logr.Logger, finalMessage, errorMessagePrefix string, writer io.Writer) ProgressLogger {
+	return &ProgressBarLogger{logger: logger, finalMessage: finalMessage, errorMessagePrefix: errorMessagePrefix, writer: writer}
+}
+
+// NewNoopProgressBar constructs a Noop Progress bar that will not display anything
+func NewNoopProgressBar(logger logr.Logger) ProgressLogger {
+	return &ProgressBarNoTTYLogger{logger: logger}
+}
+
+// ProgressBarLogger display progress bar on output
+type ProgressBarLogger struct {
```

Comment:
No unit tests on the actual progress bar

Reply:
I couldn't add unit test cases as it is causing race conditions. If we have the out as `os.stdout / os.stderr `instead of `bytes.Buffer ` , the progress bar works fine. However reading back from os.stdout to valid is challenging.